### ISSUE TITLE
Gemma 4 decode collapse: v_norm (V RMSNorm) disabled — gemma3 finding incorrectly generalized to gemma4

### ISSUE THREAD

**Issue by michalharakal** [SKaiNET-developers/SKaiNET-transformers#325]
## Summary

Gemma 4's decode via the DSL forward path (`GemmaNetworkDef`/`GemmaModel`, the only path
exercised against a live checkpoint) collapses to a degenerate output — immediate `<eos>` on
chat-template-formatted prompts, or gibberish with `GEMMA4_NO_KV_SHARING=1` — root-caused to a
**missing V-normalization step** in the attention computation.

Real Gemma 4 (`Gemma4TextAttention`, `transformers==5.15.0`) applies a parameterless
per-head RMSNorm to `V` right after `v_proj`, before it's used in attention:

```python
# Gemma4TextAttention.__init__
self.v_norm = Gemma4RMSNorm(self.head_dim, eps=config.rms_norm_eps, with_scale=False)

# Gemma4TextAttention.forward, immediately after v_proj:
value_states = self.v_norm(value_states)
value_states = value_states.transpose(1, 2)
```

`MultiHeadAttention.kt`'s own doc comment on its `vNormNoScale` parameter already correctly
describes this ("HF Gemma4TextAttention's `v_norm` is `Gemma4RMSNorm(..., with_scale=False)`"),
and the module has a working, tested implementation of exactly this normalization gated behind
that flag. But `GemmaNetworkDef.kt` passes `vNormNoScale = false` (disabled), with a comment
claiming *"gemma3 has NO v_norm ... spurious parameterless V RMS-normalization llama.cpp/HF
don't do."* That's a real finding about `gemma3` (a different, earlier architecture) that got
incorrectly generalized to `gemma4`.

## Impact

Without `v_norm`, raw (un-normalized) `V` projection output — whose per-head RMS is ~40-50 on
the real `gemma-4-E2B-it` checkpoint (confirmed matching between our loader and the real
checkpoint's weights; the *weights* were never the problem) — flows straight into the
softmax-weighted attention sum. Softmax weights themselves were verified correct (near-uniform,
not over-sharpened), so the un-normalized `V` alone produces an attention output tens of times
too large in magnitude, confirmed via a direct forward-hook comparison against the real model:

- Reference (`transformers`, real checkpoint) pre-o_proj attention output: `rms=0.952, max=13.688`
- SKaiNET (pre-fix) same point: `rms≈35-58` depending on position — **35-60× too large**

This compounds through all 35 layers into wildly wrong final logits, which is what drove the
degenerate greedy-decode symptoms.

## Repro

```
export GEMMA4_E2B_MODEL_PATH=/path/to/gemma-4-E2B-it-Q4_K_M.gguf
./gradlew :llm-runtime:kgemma:jvmTest --tests '*Gemma4ReferenceParityDiagnostic*' \
    -PuseLocalSkainet=true --rerun
```
Prints top-5 logits for the prompt `"Hi"` and a minimal chat-template prompt; before this fix,
`<eos>` wins with a large margin on any prompt containing the `<|turn>` control-token sequence.

## Fix (applied in the linked PR)

`GemmaNetworkDef.kt`: flip `vNormNoScale = false` → `true` (one line — the correct,
already-implemented normalization code path in `MultiHeadAttention.kt` just needed enabling for
Gemma 4).

## Verified impact

On the real 108-token prompt `Gemma4E2BToolCallSmokeTest` sends (system + `calculator` tool +
`"What is 17 * 23? Use the calculator tool."` + model-turn opener):
- Before: top-1 `<eos>` at `+13.67`, greedy generation immediately empty.
- After: `<eos>` no longer top-1 (`"This"`/`"The"` narrowly lead at `+18.48`/`+18.43` vs. `<eos>`
  at `+18.28`); greedy-decoded text is coherent English (`"This is a...\nThe\nThe\n..."`).

**Not a full fix** — `Gemma4E2BToolCallSmokeTest` still fails: the model degrades into a
repetition loop after ~3-4 tokens rather than reaching the exact trained `<|tool_call>...`
format. Whether that's a second, smaller bug or an inherent quantization/precision gap for this
quantized reimplementation is not yet determined — see `GEMMA4_E2B_SKAINET_FINDINGS.md` (Run 6)
for the full investigation trail and candidate next areas (`qkNorm` exactness, `gelu` formula,
final-logit-softcapping interaction).

## Scope note

Found and fixed during the same investigation as #321 (`partial_rotary_factor`) and #323
(dead-code KV-cache clobbering in the legacy `Gemma4AttentionBackend`) — those were both real,
independently-verified bugs but neither was the (sole) cause of the collapse; this is.

**Comment by michalharakal** [comment]
Closing with the parity-gate evidence the tracker asked for: the fix is merged and arch-conditional (`eee8472` enabled v_norm for gemma4, `8678265` made it gemma4-only restoring gemma3 parity), and the E2B golden-token gate now pins the outcome — `Gemma4ChatGoldenTokenTest` passes on the real checkpoint (llama.cpp b10621-pinned prompt ids + first token, packed-vs-dense A/B green), most recently re-verified through the G2 unified metadata parser (PR #383) and CI-wired into the smoke-reference tier in PR #385.
